File: python/hot100/26.环形链表2.py
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow=fast=head
        while fast and fast.next:
            fast=fast.next.next
            slow=slow.next
            if fast is slow:
                while head is not fast:
                    head=head.next
                    fast=fast.next
                return fast
        return None

File: cpp/hot100/26.环形链表2.cpp
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *slow=head,*fast=head;
        while(fast&&fast->next){
            slow=slow->next;
            fast=fast->next->next;
            if(fast==slow){
                while(slow!=head){
                    slow=slow->next;
                    head=head->next;
                }
                return slow;
            }
        }
        return NULL;
    }
};